loader.php

```diff
@@ -165,6 +163,29 @@ private function setup_actions() {
 			}, 91 );
 			add_action( 'cbox_plugins_loaded', array( 'Plugin_Dependencies', 'init' ), 91 );
 		}
+
+		// AJAX Upgrader routine.
+		add_action( 'cbox_load_components', function() {
+			// Bail if not doing AJAX.
+			if ( ! wp_doing_ajax() ) {
+				return;
+			}
+
+			// Bail if not on our special AJAX hooks.
+			if ( empty( $_POST['action'] ) ||  ( 0 !== strpos( $_POST['action'], 'cbox_' ) && false === strpos( $_POST['action'], '_upgrade' ) ) ) {
+				return;
+			}
+
+			// Ensure upgrader items are registered on AJAX.
+			$packages = cbox_get_packages();
+			$current  = cbox_get_current_package_id();
+			if ( isset( $packages[$current] ) && class_exists( $packages[$current] ) ) {
+				call_user_func( array( $packages[$current], 'upgrader' ) );
```

Comment:
Hi @r-a-y,

Based on checks above this means, that upgrades will be only registered on AJAX requests and the admin page is never visible for upgrades.

Any recommendations where should we move this logic, so upgrades are registered on admin and AJAX requests?

Reply:
Thanks for the feedback, @Mamaduka. This should be addressed in commit 9289543.

I've moved upgrader registration to the `'wp_loaded'` hook. Before, registration was too early for `cbox-openlab-core` to add their steps.